CDN/telemetry_upload.py

```diff
@@ -0,0 +1,224 @@
+    global device_key, device_secret, tm_jwt
+
+    #device_key = token_urlsafe(16)
+    device_key = ''.join(random.choice(string.ascii_letters + string.digits) for x in range(16))
+
+    # Construct a POST request to get app key and secret from reqURL
+    payload = {
+            "id": "ekstep.cdn.pinut",
+            "ver": "2.0",
+            "request": { "key": device_key}
+            }
+
+    auth_text = "bearer " + app_jwt
+    headers = {'Content-Type': 'application/json', 'Authorization': auth_text}
+
+    r  = requests.post(url=regURL, data=json.dumps(payload), headers=headers)
+    if r.status_code // 100 != 2:
+        log.error("Server error: Not received SECRET for device_key: " + device_key)
+        sys.exit(1)
+
+    device_key = r.json().get('result').get('key')
+    device_secret = r.json().get('result').get('secret')
+    #generate the telemetry jwt from app key and secret
+    tm_jwt = jwt_generate(device_key, device_secret).decode()
+    log.info("Device_key[%s] Device_secret[%s] JWT_token[%s]\n" %(device_key, device_secret, tm_jwt))
+
+
+def telemetry_upload_file(filename, jwt, endpoint=tmURL):
+    # Construct a POST request to upload telemetry
+    auth_text = "bearer " + jwt
+    headers = {'Content-Type': 'application/json', 'Content-Encoding': 'gzip', 'Authorization': auth_text}
+
+    fin = open(filename, 'rb')
+    files = {'file': fin}
+    try:
+          r  = requests.post(url=endpoint, data=fin, headers=headers)
+          print(r.text)
+    finally:
+            fin.close()
+    # Parse the response json
+
+    es_resp_status = r.json().get('params').get('status')
+    es_resp_err = r.json().get('params').get('err')
+    es_resp_errmsg = r.json().get('params').get('errmsg')
+    return (r.status_code, es_resp_status, es_resp_err, es_resp_errmsg)
+
+
+# Generate log only for 100th error from timer 
+log_optimization_limit = 100
+log_current_value = 0 
+
+def telemetry_upload_dir():
+    #
+    # Check if telemetry file avalable to sync
+    # If not, just return
+    #
+    global tmDir
+    tm_dir = tmDir
+
+    try:
+        try:
+            os.chdir(tm_dir)
+        except:
+            err_msg = "Directory read error:  " + tm_dir
+            raise BreakoutException
+
+        tmfile_list = os.listdir(tm_dir)
+        #tmfile_list = sorted(os.listdir(tm_dir),key=os.path.getctime);
+        if not tmfile_list:
+            err_msg = "No file in " + tm_dir + " to upload..."
+            raise BreakoutException
+
+        #log.info(' '.join(str(x) for x in tmfile_list))
+        #
+        # We have some files to upload; check net connectivity now
+        # If not connected, just return
+        #
+        netstatus = check_netconnectivity()
+        if not netstatus:
+            err_msg = "Not connected to network..."
+            raise BreakoutException
+        else:
+            log.info("Connected to network...")
+   
+        tmfile_timesorted_list = sorted(tmfile_list, key=os.path.getmtime)
```

Comment:
sorted gives oldest file first ?  We need to send oldest file first ?

Reply:
Yes, first in first out.